pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/DataBuffer.java

```diff
@@ -0,0 +1,292 @@
+   * Implementations should do their best to do not allocate memory and instead return a view of the underlying data.
+   * Callers cannot assume they have the ownership of the returned ByteBuffer, and therefore should not call
+   * {@link CleanerUtil#cleanQuietly(ByteBuffer)} using the returned object.
+   */
+  ByteBuffer copyOrView(long offset, int size, ByteOrder byteOrder);
+
+  void appendAsByteBuffers(List<ByteBuffer> appendTo);
+
+  /**
+   *
+   * Returns a ByteBuffer whose content is the same as the range [offset, offset + size) of this buffer.
+   * <p>
+   * Implementations should do their best to do not allocate memory and instead return a view of the underlying data.
+   * Callers cannot assume they have the ownership of the returned ByteBuffer, and therefore should not call
+   * {@link CleanerUtil#cleanQuietly(ByteBuffer)} using the returned object.
+   * <p>
+   * The returned ByteBuffer will have the same byte order as this buffer.
+   */
+  default ByteBuffer copyOrView(long offset, int size) {
+    return copyOrView(offset, size, order());
+  }
+
+  static boolean sameContent(DataBuffer buffer1, DataBuffer buffer2) {
+    if (buffer1.size() != buffer2.size()) {
+      return false;
+    }
+    DataBuffer nativeBuffer1 = buffer1.view(0, buffer1.size(), ByteOrder.nativeOrder());
+    DataBuffer nativeBuffer2 = buffer2.view(0, buffer1.size(), ByteOrder.nativeOrder());
+
+    long maxLong = nativeBuffer1.size() & ~7L;
+    for (long i = 0; i < maxLong; i += 8) {
+      if (nativeBuffer1.getLong(i) != nativeBuffer2.getLong(i)) {
+        return false;
+      }
+    }
+    for (long i = maxLong; i < buffer1.size(); i++) {
+      if (buffer1.getByte(i) != buffer2.getByte(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  static int commonHashCode(DataBuffer dataBuffer) {
+    long firstLong;
+    long lastLong;
+    long size = dataBuffer.size();
+    int intSize;
+    if (size > Integer.MAX_VALUE) {
+      intSize = Integer.MAX_VALUE;
+    } else {
+      intSize = (int) size;
+    }
+    if (intSize >= 8) {
+      firstLong = dataBuffer.getLong(0);
+      lastLong = dataBuffer.getLong(size - 8);
+    } else {
+      switch (intSize) {
+        case 0:
+          firstLong = 0;
+          lastLong = 0;
+          break;
+        case 1:
+          firstLong = dataBuffer.getByte(0);
+          lastLong = firstLong;
+          break;
+        case 2:
+          firstLong = dataBuffer.getShort(0);
+          lastLong = firstLong;
+          break;
+        case 3:
+          firstLong = dataBuffer.getShort(0);
+          lastLong = dataBuffer.getShort(1);
+          break;
+        case 4:
+          firstLong = dataBuffer.getInt(0);
+          lastLong = firstLong;
+          break;
+        case 5:
+        case 6:
+        case 7:
+          firstLong = dataBuffer.getInt(0);
+          lastLong = dataBuffer.getInt(intSize - 4);
+          break;
+        default:
+          firstLong = dataBuffer.getLong(0);
```

Comment:
(minor) not reachable, or consider removing the `if` check above

Reply:
Done